submit_cgap/tests/test_submission.py

```diff
@@ -2202,3 +2253,144 @@ def test_running_on_windows_native():
         os_name, is_windows = pair
         with mock.patch.object(os, "name", os_name):
             assert running_on_windows_native() is is_windows
+
+
+@pytest.mark.parametrize(
+    "directory,file_name,recursive,glob_results,expected_file_path,expected_msg",
+    [
+        ("foo", "bar", False, [], "foo/bar", False),
+        ("foo", "bar", True, [], "foo/bar", False),
+        ("foo", "bar", False, ["foo/bar"], "foo/bar", False),
+        ("foo", "bar", False, ["foo/bar", "fu/foo/bar"], None, True),
+    ]
+)
+def test_search_for_file(
+    directory, file_name, recursive, glob_results, expected_file_path, expected_msg
+):
+    """Test output file path +/- error message dependent on file search
+    via glob.
+    """
+    with mock.patch(
+        "submit_cgap.submission.glob.glob", return_value=glob_results
+    ) as mocked_glob:
+        file_path_found, error_msg = search_for_file(directory, file_name, recursive)
+        mocked_glob.assert_called_once_with(
+            directory + "/" + file_name, recursive=recursive
+        )
+        assert file_path_found == expected_file_path
+        if expected_msg:
+            assert error_msg.startswith(
+                f"No upload attempted for file {file_name}"
+            )
+
+
+@pytest.mark.parametrize(
+    "no_query,cgap_selective_uploads,yes_or_no_result,error_raised,expected_result",
+    [
+        (False, True, False, None, None),
+        (False, False, False, None, None),
+        (False, True, True, None, "something"),
+        (False, True, True, True, None),
+        (True, True, False, None, "something"),
+        (True, True, False, True, None),
+    ]
+)
+def test_wrap_upload_function(
+    no_query, cgap_selective_uploads, yes_or_no_result, error_raised, expected_result
+):
+    """Test UploadMessageWrapper.wrap_upload_function creates
+    appropriate messages on given upload function.
+
+    Ensure wrapped function is indeed wrapped and returns expected
+    output when called.
+    """
+    with shown_output() as shown:
+        with mock.patch.object(
+            submission_module, "yes_or_no", return_value=yes_or_no_result
+        ) as mocked_yes_or_no:
+            with mock.patch.object(
+                submission_module,
+                "CGAP_SELECTIVE_UPLOADS",
+                cgap_selective_uploads,
+            ):
+                side_effect = None
+                if error_raised:
+                    side_effect = RuntimeError("Error occurred")
+                simple_function = mock.MagicMock(
+                    side_effect=side_effect, return_value=expected_result
+                )
+
+                uuid = "some_uuid"
+                file_name = "foo/bar"
+                input_arg = "foo"
+                function_wrapper = UploadMessageWrapper(uuid, no_query=no_query)
+                result = function_wrapper.wrap_upload_function(
+                    simple_function, file_name
+                )(input_arg, error_raised=error_raised)
```

Comment:
I have a preference for going in steps both because it's WAY easier if you actually have to step things (or otherwise you end up stepping down into the same line twice, and the controls are not always intelligible, depending on your interface) and also because I think it's more readable because the syntax is more natural. So I'd write (untested):
```suggestion
                wrapped_uploader = function_wrapper.wrap_upload_function(simple_function,
                                                                         file_name)
                result = wrapped_uploader(input_arg, error_raised=error_raised)
```